### PR TITLE
Added ping/echo to lampshade protocol

### PR DESCRIPTION
**Pull request by oxtoacart**
@myleshorton @joesis 

This adds a ping/echo protocol to lampshade that supports periodic RTT timing. Ping frames only go out attached to other frames, so on a completely idle connection there is no additional overhead.

The RTT timings will be useful for proxy selection.

### CODE REVIEW

File: lampshade.go [hunk 1]
```diff
@@ -123,6 +126,9 @@
 //
 //     Data     - data (for type "data" or "padding")
 //
+//     TS       - time at which ping packet was sent as 64-bit uint
```
Comment (oxtoacart) [inline lampshade.go]: Fixed.

File: session.go [hunk 2]
```diff
@@ -18,40 +22,51 @@ type session struct {
...
+// message will be sent with the first frame sent in this session.
+func startSession(conn net.Conn, windowSize int, maxPadding int, pingInterval time.Duration, decrypt func([]byte), encrypt func(dst []byte, src []byte), clientInitMsg []byte, pool BufferPool, connCh chan net.Conn, beforeClose func(*session)) *session {
 	s := &session{
 		Conn:          conn,
 		windowSize:    windowSize,
 		maxPadding:    big.NewInt(int64(maxPadding)),
 		decrypt:       decrypt,
 		encrypt:       encrypt,
 		clientInitMsg: clientInitMsg,
+		isClient:      clientInitMsg != nil,
```
Comment (oxtoacart) [inline session.go]: Fixed.

**Comment by oxtoacart** [comment]
@joesis Thanks for reviewing! I've addressed your comments and also made a small performance tweak related to use of timers.

File: lampshade.go [hunk 3]
```diff
@@ -328,3 +366,18 @@ func ackWithFrames(header []byte, frames int32) []byte {
 	binaryEncoding.PutUint32(ack, uint32(frames))
 	return ack
 }
+
+func ping() []byte {
+	// note - header and ts field are reversed to match the usual format for
+	// data frames
+	ping := make([]byte, pingFrameSize)
+	ping[tsSize] = frameTypePing
+	binaryEncoding.PutUint64(ping, uint64(mtime.Now()))
```
Comment (joesis) [inline lampshade.go]: One last comment: the `TS` is not nanoseconds since epoch!

Comment (oxtoacart) [inline lampshade.go]: Oh, haha, good point. Will update comment.

Comment (oxtoacart) [inline lampshade.go]: Fixed.

**Comment by joesis** [comment]
Cool. LGTM

**Pull request merged**